go-controller/pkg/ovn/service.go

```diff
@@ -191,117 +165,105 @@ func (ovn *Controller) createService(service *kapi.Service) error {
 					continue
 				}
 				for _, physicalIP := range physicalIPs {
 					// With the physical_ip:port as the VIP, add an entry in
-					// 'load_balancer'.
+					// 'load balancer'.
 					vip := util.JoinHostPortInt32(physicalIP, port)
 					// Skip creating LB if endpoints watcher already did it
 					if _, hasEps := ovn.getServiceLBInfo(loadBalancer, vip); hasEps {
-						klog.V(5).Infof("Load Balancer already configured for %s, %s", loadBalancer, vip)
+						klog.V(5).Infof("Load balancer already configured for %s, %s", loadBalancer, vip)
 					} else if ep != nil {
 						if err := ovn.AddEndpoints(ep); err != nil {
 							return err
 						}
 					} else if ovn.svcQualifiesForReject(service) {
-						aclUUID, err := ovn.createLoadBalancerRejectACL(loadBalancer, physicalIP, port, protocol)
+						aclUUID, err := ovn.createLoadBalancerRejectACL(loadBalancer, physicalIP, port, svcPort.Protocol)
 						if err != nil {
 							return fmt.Errorf("failed to create service ACL: %v", err)
 						}
-						klog.V(5).Infof("Service Reject ACL created for physical gateway: %s", aclUUID)
+						klog.V(5).Infof("Service Reject ACL created for gateway router: %s", aclUUID)
 					}
 				}
 			}
 		}
 		if util.ServiceTypeHasClusterIP(service) {
-			loadBalancer, err := ovn.getLoadBalancer(protocol)
+			loadBalancer, err := ovn.getLoadBalancer(svcPort.Protocol)
 			if err != nil {
-				klog.Errorf("Failed to get load-balancer for %s (%v)",
-					protocol, err)
+				klog.Errorf("Failed to get load balancer for %s (%v)", svcPort.Protocol, err)
 				break
 			}
 			if ovn.svcQualifiesForReject(service) {
 				vip := util.JoinHostPortInt32(service.Spec.ClusterIP, svcPort.Port)
 				// Skip creating LB if endpoints watcher already did it
 				if _, hasEps := ovn.getServiceLBInfo(loadBalancer, vip); hasEps {
-					klog.V(5).Infof("Load Balancer already configured for %s, %s", loadBalancer, vip)
+					klog.V(5).Infof("Load balancer already configured for %s, %s", loadBalancer, vip)
 				} else if ep != nil {
 					if err := ovn.AddEndpoints(ep); err != nil {
 						return err
 					}
 				} else {
 					aclUUID, err := ovn.createLoadBalancerRejectACL(loadBalancer, service.Spec.ClusterIP,
-						svcPort.Port, protocol)
+						svcPort.Port, svcPort.Protocol)
 					if err != nil {
 						return fmt.Errorf("failed to create service ACL: %v", err)
-					} else {
-						klog.V(5).Infof("Service Reject ACL created for cluster IP: %s", aclUUID)
 					}
+					klog.V(5).Infof("Service Reject ACL created for cluster IP: %s", aclUUID)
 				}
 				for _, extIP := range service.Spec.ExternalIPs {
-					exLoadBalancer := ovn.getDefaultGatewayLoadBalancer(svcPort.Protocol)
-					if exLoadBalancer == "" {
-						klog.Warningf("No default gateway found for protocol %s\n\tNote: 'nodeport'"+
-							"flag needs to be enabled for default gateway", svcPort.Protocol)
-						continue
+					gateways, _, err := ovn.getOvnGateways()
+					if err != nil {
+						return err
 					}
```

Comment:
L248-250: Can you please call this once before L247, if len(service.Spec.ExternalIPs) > 0, since it is expensive,  say in 500-node K8s cluster.